### PR TITLE
feat: fetchMarginAdjustmentHistory

### PR DESCRIPTION
**Pull request by samgermain**
## binance

```
Python v3.9.6
CCXT v4.2.80
binance.fetchMarginAdjustmentHistory(XRP/USDT:USDT)
[{'amount': 2.5714824,
  'code': 'USDT',
  'datetime': '2024-03-21T18:37:51.555Z',
  'info': {'amount': '2.57148240',
           'asset': 'USDT',
           'clientTranId': '',
           'deltaType': 'TRADE',
           'positionSide': 'BOTH',
           'symbol': 'XRPUSDT',
           'time': '1711046271555',
           'type': '1'},
  'status': 'ok',
  'symbol': 'XRP/USDT:USDT',
  'timestamp': 1711046271555,
  'type': 'add'},
 ...
 ]
```

```
Python v3.9.6
CCXT v4.2.80
binance.fetchMarginAdjustmentHistory(XRP/USDT:USDT,undefined,1710460800000,3,{'until': '1711146090000'})
[{'amount': -0.07212375,
  'code': 'USDT',
  'datetime': '2024-03-22T00:00:18.176Z',
  'info': {'amount': '-0.07212375',
           'asset': 'USDT',
           'clientTranId': '',
           'deltaType': 'FUNDING_FEE',
           'positionSide': 'BOTH',
           'symbol': 'XRPUSDT',
           'time': '1711065618176',
           'type': '2'},
  'status': 'ok',
  'symbol': 'XRP/USDT:USDT',
  'timestamp': 1711065618176,
  'type': 'reduce'},
 ...
 ]

```

```
% py binance addMargin XRP/USDT:USDT 10 --testnet                         
Python v3.9.6
CCXT v4.2.80
binance.addMargin(XRP/USDT:USDT,10)
{'amount': 10.0,
 'code': 'USDT',
 'datetime': None,
 'info': {'amount': '10',
          'code': '200',
          'msg': 'Successfully modify position margin.',
          'type': '1'},
 'status': 'ok',
 'symbol': 'XRP/USDT:USDT',
 'timestamp': None,
 'type': 'add'}
 ```

 ```
 % py binance reduceMargin XRP/USDT:USDT 10 --testnet
Python v3.9.6
CCXT v4.2.82
binance.reduceMargin(XRP/USDT:USDT,10)
{'amount': 10.0,
 'code': 'USDT',
 'datetime': None,
 'info': {'amount': '10',
          'code': '200',
          'msg': 'Successfully modify position margin.',
          'type': '2'},
 'status': 'ok',
 'symbol': 'XRP/USDT:USDT',
 'timestamp': None,
 'type': 'reduce'}
 ```

 ## coinex

```
 Python v3.9.6
CCXT v4.2.80
coinex.fetchMarginAdjustmentHistory(XRP/USDT:USDT)
[{'amount': 3.0,
  'code': 'USDT',
  'datetime': '2024-03-21T19:55:06.382Z',
  'info': {'bkr_price': '0',
           'leverage': '3',
           'liq_price': '0',
           'margin_amount': '5.33236666666666666666',
           'margin_change': '3',
           'market': 'XRPUSDT',
           'position_amount': '11',
           'position_id': '297155652',
           'position_type': '2',
           'settle_price': '0.6361',
           'time': '1711050906.382891',
           'type': '1',
           'user_id': '3685860'},
  'status': None,
  'symbol': 'XRP/USDT:USDT',
  'timestamp': 1711050906382,
  'type': 'add'}]
```

```
% py coinex addMargin XRP/USDT:USDT 3           
Python v3.9.6
CCXT v4.2.80
coinex.addMargin(XRP/USDT:USDT,3)
{'amount': 3.0,
 'code': 'USDT',
 'datetime': '2024-03-22T23:35:31.978Z',
 'info': {'adl_level': '1',
          'adl_sort': '1386',
          'adl_sort_val': '0.99869776',
          'amount': '10',
          'amount_max': '10',
          'amount_max_margin': '2.03466666666666666666',
          'auction_start_amount': '0.00000000000000000000',
          'bkr_price': '0.00000000000000000000',
          'bkr_price_imply': '0.10693333333333333333',
          'close_left': '10',
          'create_time': '1711150526.2581',
          'deal_all': '6.10400000000000000000',
          'deal_asset_fee': '0.00000000000000000000',
          'fee_asset': '',
          'finish_type': '1',
          'first_price': '0.6104',
          'insurance': '0.00000000000000000000',
          'latest_price': '0.6104',
          'leverage': '3',
          'liq_amount': '0.00000000000000000000',
          'liq_order_price': '0',
          'liq_order_time': '0',
          'liq_price': '0.00000000000000000000',
          'liq_price_imply': '0.10801346801346801347',
          'liq_profit': '0.00000000000000000000',
          'liq_time': '0',
          'mainten_margin': '0.01',
          'mainten_margin_amount': '0.06096000000000000000',
          'maker_fee': '0.00000000000000000000',
          'margin_amount': '5.03466666666666666666',
          'market': 'XRPUSDT',
          'open_margin': '1.64040618500504914809',
          'open_margin_imply': '0.82481432940148536478',
          'open_price': '0.61040000000000000000',
          'open_val': '6.10400000000000000000',
          'open_val_max': '6.10400000000000000000',
          'position_id': '297371462',
          'profit_clearing': '-0.00305200000000000000',
          'profit_real': '-0.00305200000000000000',
          'profit_unreal': '-0.0080',
          'settle_price': '0.61040000000000000000',
          'settle_val': '6.10400000000000000000',
          'side': '2',
          'stop_loss_price': '0.00000000000000000000',
          'stop_loss_type': '0',
          'sys': '0',
          'take_profit_price': '0.00000000000000000000',
          'take_profit_type': '0',
          'taker_fee': '0.00000000000000000000',
          'total': '1590',
          'type': '2',
          'update_time': '1711150531.978157',
          'user_id': '3685860'},
 'status': 'OK',
 'symbol': 'XRP/USDT:USDT',
 'timestamp': 1711150531978,
 'type': 'reduce'}
```

## okx

```
Python v3.9.6
CCXT v4.2.81
okx.fetchMarginAdjustmentHistory(XRP/USDT:USDT,add)
[{'amount': 10.0,
  'code': 'USDT',
  'datetime': '2024-03-22T07:21:17.132Z',
  'info': {'bal': '67561.4325135010619812',
           'balChg': '-10.0000000000000000',
           'billId': '691305508795404288',
           'ccy': 'USDT',
           'clOrdId': '',
           'execType': '',
           'fee': '0',
           'fillFwdPx': '',
           'fillIdxPx': '',
           'fillMarkPx': '',
           'fillMarkVol': '',
           'fillPxUsd': '',
           'fillPxVol': '',
           'fillTime': '1711092077283',
           'from': '',
           'instId': 'XRP-USDT-SWAP',
           'instType': 'SWAP',
           'interest': '0',
           'mgnMode': 'isolated',
           'notes': '',
           'ordId': '',
           'pnl': '0',
           'posBal': '133.12',
           'posBalChg': '10.00',
           'px': '',
           'subType': '160',
           'sz': '10',
           'tag': '',
           'to': '',
           'tradeId': '0',
           'ts': '1711092077132',
           'type': '6'},
  'status': None,
  'symbol': 'XRP/USDT:USDT',
  'timestamp': 1711092077132,
  'type': 'add'},
  ...
]
```

```
Python v3.9.6
CCXT v4.2.81
okx.fetchMarginAdjustmentHistory(XRP/USDT:USDT,reduce)
[{'amount': 10.0,
  'code': 'USDT',
  'datetime': '2024-03-22T07:03:07.831Z',
  'info': {'bal': '67581.4325135010619812',
           'balChg': '10.0000000000000000',
           'billId': '691300939935862784',
           'ccy': 'USDT',
           'clOrdId': '',
           'execType': '',
           'fee': '0',
           'fillFwdPx': '',
           'fillIdxPx': '',
           'fillMarkPx': '',
           'fillMarkVol': '',
           'fillPxUsd': '',
           'fillPxVol': '',
           'fillTime': '1711090988014',
           'from': '',
           'instId': 'XRP-USDT-SWAP',
           'instType': 'SWAP',
           'interest': '0',
           'mgnMode': 'isolated',
           'notes': '',
           'ordId': '',
           'pnl': '0',
           'posBal': '113.12',
           'posBalChg': '-10.00',
           'px': '',
           'subType': '161',
           'sz': '10',
           'tag': '',
           'to': '',
           'tradeId': '0',
           'ts': '1711090987831',
           'type': '6'},
  'status': None,
  'symbol': 'XRP/USDT:USDT',
  'timestamp': 1711090987831,
  'type': 'reduce'}]
```

### CODE REVIEW

**Comment by carlosmiei** [comment]
@samgermain can you check the bill and add docs for this method?

**Comment by carlosmiei** [comment]
@samgermain can you rebase this PR?

**Comment by samgermain** [comment]
I'm getting these errors when running tests, which seem like they're unrelated

```
C# version: 7.0.17
[TEST_FAILURE][C#][STATIC_REQUEST_TEST_FAILURE][coinex  spot [subType: linear] ][fetchTime][Fetch the time]System.Exception: Assertion failed: url mismatch computed "/perpetual/v1/time" stored: "/v2/time"
   at Tests.BaseTest.Assert(Object condition2, Object message2) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/BaseTest.Bridge.cs:line 54
   at Tests.BaseTest.assert(Object condition, Object message) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/BaseTest.Bridge.cs:line 58
   at Tests.testMainClass.assertStaticError(Object cond, Object message, Object calculatedOutput, Object storedOutput, Object key) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/Generated/TestMethods.cs:line 866
   at Tests.testMainClass.assertStaticRequestOutput(Exchange exchange, Object type, Object skipKeys, Object storedUrl, Object requestUrl, Object storedOutput, Object newOutput) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/Generated/TestMethods.cs:line 1084
   at Tests.testMainClass.testMethodStatically(Exchange exchange, Object method, Object data, Object type, Object skipKeys) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/Generated/TestMethods.cs:line 1183
[TEST_FAILURE][C#][STATIC_REQUEST_TEST_FAILURE][coinex  spot [subType: linear] ][fetchTradingFee][Spot fetch trading fee]System.Exception: Assertion failed: url mismatch computed "/v1/market/detail" stored: "/v2/spot/market"
   at Tests.BaseTest.Assert(Object condition2, Object message2) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/BaseTest.Bridge.cs:line 54
   at Tests.BaseTest.assert(Object condition, Object message) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/BaseTest.Bridge.cs:line 58
   at Tests.testMainClass.assertStaticError(Object cond, Object message, Object calculatedOutput, Object storedOutput, Object key) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/Generated/TestMethods.cs:line 866
   at Tests.testMainClass.assertStaticRequestOutput(Exchange exchange, Object type, Object skipKeys, Object storedUrl, Object requestUrl, Object storedOutput, Object newOutput) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/Generated/TestMethods.cs:line 1084
   at Tests.testMainClass.testMethodStatically(Exchange exchange, Object method, Object data, Object type, Object skipKeys) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/Generated/TestMethods.cs:line 1183
[TEST_FAILURE][C#][STATIC_REQUEST_TEST_FAILURE][coinex  spot [subType: linear] ][fetchTradingFee][Swap fetch trading fee]System.Exception: Assertion failed: url mismatch computed "/v1/market/detail" stored: "/v2/futures/market"
   at Tests.BaseTest.Assert(Object condition2, Object message2) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/BaseTest.Bridge.cs:line 54
   at Tests.BaseTest.assert(Object condition, Object message) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/BaseTest.Bridge.cs:line 58
   at Tests.testMainClass.assertStaticError(Object cond, Object message, Object calculatedOutput, Object storedOutput, Object key) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/Generated/TestMethods.cs:line 866
   at Tests.testMainClass.assertStaticRequestOutput(Exchange exchange, Object type, Object skipKeys, Object storedUrl, Object requestUrl, Object storedOutput, Object newOutput) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/Generated/TestMethods.cs:line 1084
   at Tests.testMainClass.testMethodStatically(Exchange exchange, Object method, Object data, Object type, Object skipKeys) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/Generated/TestMethods.cs:line 1183
[TEST_FAILURE][C#][STATIC_REQUEST_TEST_FAILURE][coinex  spot [subType: linear] ][fetchTradingFees][Spot fetch trading fees]System.Exception: Assertion failed: url mismatch computed "/v1/market/info" stored: "/v2/spot/market"
   at Tests.BaseTest.Assert(Object condition2, Object message2) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/BaseTest.Bridge.cs:line 54
   at Tests.BaseTest.assert(Object condition, Object message) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/BaseTest.Bridge.cs:line 58
   at Tests.testMainClass.assertStaticError(Object cond, Object message, Object calculatedOutput, Object storedOutput, Object key) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/Generated/TestMethods.cs:line 866
   at Tests.testMainClass.assertStaticRequestOutput(Exchange exchange, Object type, Object skipKeys, Object storedUrl, Object requestUrl, Object storedOutput, Object newOutput) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/Generated/TestMethods.cs:line 1084
   at Tests.testMainClass.testMethodStatically(Exchange exchange, Object method, Object data, Object type, Object skipKeys) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/Generated/TestMethods.cs:line 1183
[TEST_FAILURE][C#][STATIC_REQUEST_TEST_FAILURE][coinex  spot [subType: linear] ][fetchTradingFees][Swap fetch trading fees]System.Exception: Assertion failed: url mismatch computed "/v1/market/info" stored: "/v2/futures/market"
   at Tests.BaseTest.Assert(Object condition2, Object message2) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/BaseTest.Bridge.cs:line 54
   at Tests.BaseTest.assert(Object condition, Object message) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/BaseTest.Bridge.cs:line 58
   at Tests.testMainClass.assertStaticError(Object cond, Object message, Object calculatedOutput, Object storedOutput, Object key) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/Generated/TestMethods.cs:line 866
   at Tests.testMainClass.assertStaticRequestOutput(Exchange exchange, Object type, Object skipKeys, Object storedUrl, Object requestUrl, Object storedOutput, Object newOutput) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/Generated/TestMethods.cs:line 1084
   at Tests.testMainClass.testMethodStatically(Exchange exchange, Object method, Object data, Object type, Object skipKeys) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/Generated/TestMethods.cs:line 1183
Unhandled exception. System.AggregateException: One or more errors occurred. (Object reference not set to an instance of an object.)
 ---> System.NullReferenceException: Object reference not set to an instance of an object.
   at Tests.testMainClass.testMethodStatically(Exchange exchange, Object method, Object data, Object type, Object skipKeys) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/Generated/TestMethods.cs:line 1175
   at Tests.testMainClass.testExchangeRequestStatically(Object exchangeName, Object exchangeData, Object testName) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/Generated/TestMethods.cs:line 1278
   at ccxt.Exchange.PromiseAll(Object promisesObj) in /Users/samgermain/Documents/CCXT/ccxt/cs/ccxt/base/Exchange.TranspileHelpers.cs:line 808
   at Tests.testMainClass.runStaticTests(Object type, Object targetExchange, Object testName) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/Generated/TestMethods.cs:line 1396
   at Tests.testMainClass.runStaticRequestTests(Object targetExchange, Object testName) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/Generated/TestMethods.cs:line 1359
   at Tests.testMainClass.init(Object exchangeId, Object symbolArgv) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/Generated/TestMethods.cs:line 40
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Tests.Tests.Main(String[] args) in /Users/samgermain/Documents/CCXT/ccxt/cs/tests/Program.cs:line 120
   ```

**Comment by carlosmiei** [comment]
@samgermain yes those errors are unrelated

**Comment by carlosmiei** [comment]
@samgermain can you please add this new method to the manual?

File: ts/src/base/Exchange.ts [hunk 1]
```diff
@@ -2404,6 +2405,21 @@ export default class Exchange {
         throw new NotSupported (this.id + ' setMargin() is not supported yet');
     }
 
+    async fetchMarginAdjustmentHistory (symbol: Str = undefined, type: Str = undefined, since: Num = undefined, limit: Num = undefined, params = {}): Promise<object[]> {
```
Comment (carlosmiei) [inline ts/src/base/Exchange.ts:2408]: @samgermain this return type will not work properly in c#, can you add a proper type for this method?

**Pull request merged**